File: DynamicLights/src/main/java/atomicstryker/dynamiclights/common/DLTransformer.java
package atomicstryker.dynamiclights.common;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Iterator;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.VarInsnNode;
import org.objectweb.asm.util.Printer;
import org.objectweb.asm.util.Textifier;
import org.objectweb.asm.util.TraceMethodVisitor;

import net.minecraft.launchwrapper.IClassTransformer;

/**
 * 
 * @author AtomicStryker
 * 
 * Magic happens in here. MAGIC.
 * Obfuscated names will have to be updated with each Obfuscation change.
 *
 */
public class DLTransformer implements IClassTransformer
{
    
    /* net/minecraft/world/World */
    private String classNameWorld = "amu";

    /*
     * (Lnet/minecraft/util/BlockPos;Lnet/minecraft/world/EnumSkyBlock;)I / func_175638_a
     */
    private String targetMethodDesc = "(Let;Lana;)I";

    /* net/minecraft/world/World.getRawLight / func_175638_a */
    private String computeLightValueMethodName = "a";

    /*
     * (Lnet/minecraft/block/Block;Lnet/minecraft/block/state/IBlockState;Lnet/minecraft/world/IBlockAccess;Lnet/minecraft/util/BlockPos;)I
     */
    private String goalInvokeDesc = "(Laow;Lawt;Lamy;Let;)I";
    
    @Override
    public byte[] transform(String name, String newName, byte[] bytes)
    {
        //System.out.println("transforming: "+name);
        if (name.equals(classNameWorld))
        {
            return handleWorldTransform(bytes, true);
        }
        else if (name.equals("net.minecraft.world.World")) // MCP testing
        {
            computeLightValueMethodName = "getRawLight";
            goalInvokeDesc = "(Lnet/minecraft/block/Block;Lnet/minecraft/block/state/IBlockState;Lnet/minecraft/world/IBlockAccess;Lnet/minecraft/util/math/BlockPos;)I";
            return handleWorldTransform(bytes, false);
        }
        
        return bytes;
    }
    
    private String insnToString(AbstractInsnNode insn)
    {
        insn.accept(mp);
        StringWriter sw = new StringWriter();
        printer.print(new PrintWriter(sw));
        printer.getText().clear();
        return sw.toString();
    }

    private Printer printer = new Textifier();
    private TraceMethodVisitor mp = new TraceMethodVisitor(printer); 
    
    private byte[] handleWorldTransform(byte[] bytes, boolean obf)
    {
        System.out.println("**************** Dynamic Lights transform running on World, obf: "+obf+" *********************** ");
        ClassNode classNode = new ClassNode();
        ClassReader classReader = new ClassReader(bytes);
        classReader.accept(classNode, 0);
        
        // find method to inject into
        for (MethodNode m : classNode.methods)
        {
            if (m.name.equals(computeLightValueMethodName)
                    && (!obf || m.desc.equals(targetMethodDesc)))
            {
                System.out.println("In target method " + computeLightValueMethodName +":"+m.desc+ ", Patching!");
                
                /* before patch, forge 2698
                   
               aload_2
               getstatic     #136               // Field net/minecraft/world/EnumSkyBlock.SKY:Lnet/minecraft/world/EnumSkyBlock;
               if_acmpne     18
               aload_0       
               aload_1       
               invokevirtual #160               // Method isAgainstSky:(Lnet/minecraft/util/BlockPos;)Z
               ifeq          18
               bipush        15
               ireturn       
               aload_0       
               aload_1       
               invokevirtual #80                // Method getBlockState:(Lnet/minecraft/util/BlockPos;)Lnet/minecraft/block/state/IBlockState;
               astore_3       
               getstatic                        // Field net/minecraft/world/EnumSkyBlock.SKY:Lnet/minecraft/world/EnumSkyBlock;
               if_acmpne
               iconst_0
               goto [control jumps somewhere else]
               aload_3
               invokeinterface                  // InterfaceMethod Lnet/minecraft/block/state/IBlockState.getBlock:()Lnet/minecraft/block/Block;
               aload_3
               aload_0       
               aload_1       
               stack is block, iblockstate, iblockaccess, blockpos
               invokevirtual                    // Method net/minecraft/block/Block.getLightValue (Lnet/minecraft/block/state/IBlockState;Lnet/minecraft/world/IBlockAccess;Lnet/minecraft/util/BlockPos;)I
               istore        4
              
               [... many more...]
              */
                
                /*
                System.out.println("=== PRE PRINT===");
                Iterator<AbstractInsnNode> printIter = m.instructions.iterator();
                while (printIter.hasNext())
                {
                    System.out.print(insnToString(printIter.next()));
                }
                System.out.println();
                System.out.println("=== PRE PRINT===");
                */

                AbstractInsnNode targetNode = null;
                Iterator<AbstractInsnNode> iter = m.instructions.iterator();
                boolean found = false;
                int index = 0;
                while (iter.hasNext())
                {
                    // check all nodes
                    targetNode = iter.next();

                    // find the first ASTORE node, it stores the Block reference for the Block.getLightValue call
                    if (targetNode.getOpcode() == Opcodes.ASTORE)
                    {
                        VarInsnNode astore = (VarInsnNode) targetNode;
                        System.out.println("Found ASTORE Node at index "+index+", is writing variable number: " + astore.var);

                        // go further until ISTORE is hit
                        while (targetNode.getOpcode() != Opcodes.ISTORE)
                        {
                            // we want to delete the INVOKEVIRTUAL aka the call to Block.getLightValue
                            if (targetNode instanceof MethodInsnNode && targetNode.getOpcode() == Opcodes.INVOKEVIRTUAL)
                            {
                                MethodInsnNode mNode = (MethodInsnNode) targetNode;
                                System.out.printf("found deletion target at index %d: %s\n", index, insnToString(mNode));
                                found = true;
                                iter.remove();
                                targetNode = iter.next(); // select next node as target
                                break;
                            }
                            targetNode = iter.next();
                            System.out.print("Reading node: "+insnToString(targetNode));
                        }
                        break;
                    }
                    index++;
                }

                if (found)
                {
                    // now write our replacement before the target node
                    m.instructions.insertBefore(targetNode, new MethodInsnNode(Opcodes.INVOKESTATIC, "atomicstryker/dynamiclights/client/DynamicLights", "getLightValue", goalInvokeDesc, false));
                }
                break;
            }
        }
        
        ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_MAXS|ClassWriter.COMPUTE_FRAMES);
        classNode.accept(writer);
        return writer.toByteArray();
    }
}

File: DynamicLights/src/main/resources/META-INF/dltransformer.js
function initializeCoreMod() {
    print("Dynamic Lights cminit");

    var/*Class*/ ASMAPI = Java.type('net.minecraftforge.coremod.api.ASMAPI');

    var/*Class*/ VarInsnNode = Java.type('org.objectweb.asm.tree.VarInsnNode');
    var/*Class*/ MethodInsnNode = Java.type('org.objectweb.asm.tree.MethodInsnNode');
    var/*Class*/ InsnList = Java.type('org.objectweb.asm.tree.InsnList');

    var/*Class/Interface*/ Opcodes = Java.type('org.objectweb.asm.Opcodes');

    // var methodName = "getLightFor";
    var methodName = ASMAPI.mapMethod("func_175642_b");
    print("func_175642_b was mapped to: ", methodName);

    var className = 'net.minecraft.world.World';
    var methodDescriptorToModify = "(Lnet/minecraft/world/LightType;Lnet/minecraft/util/math/BlockPos;)I";
    var goalInvokeDesc = "(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;I)I";

    if (!methodName.equals("getLightFor")) {
        print("detecting obfuscated environment, method name is ", methodName);
    }

    /*
          public getLightFor(Lnet/minecraft/world/LightType;Lnet/minecraft/util/math/BlockPos;)I
           L0
            LINENUMBER 416 L0
            ALOAD 0
            INVOKEVIRTUAL net/minecraft/world/World.getChunkProvider ()Lnet/minecraft/world/chunk/AbstractChunkProvider;
            INVOKEVIRTUAL net/minecraft/world/chunk/AbstractChunkProvider.getLightManager ()Lnet/minecraft/world/lighting/WorldLightManager;
            ALOAD 1
            INVOKEVIRTUAL net/minecraft/world/lighting/WorldLightManager.getLightEngine (Lnet/minecraft/world/LightType;)Lnet/minecraft/world/lighting/IWorldLightListener;
            ALOAD 2
            INVOKEINTERFACE net/minecraft/world/lighting/IWorldLightListener.getLightFor (Lnet/minecraft/util/math/BlockPos;)I (itf)
            IRETURN
           L1
            LOCALVARIABLE this Lnet/minecraft/world/World; L0 L1 0
            LOCALVARIABLE type Lnet/minecraft/world/LightType; L0 L1 1
            LOCALVARIABLE pos Lnet/minecraft/util/math/BlockPos; L0 L1 2
            MAXSTACK = 2
            MAXLOCALS = 3

            transform to:

          public getLightFor(Lnet/minecraft/world/LightType;Lnet/minecraft/util/math/BlockPos;)I
           L0
            LINENUMBER 112 L0
            ALOAD 0
            INVOKEVIRTUAL atomicstryker/dynamiclights/client/WorldTest.getChunkProvider ()Lnet/minecraft/world/chunk/AbstractChunkProvider;
            INVOKEVIRTUAL net/minecraft/world/chunk/AbstractChunkProvider.getLightManager ()Lnet/minecraft/world/lighting/WorldLightManager;
            ALOAD 1
            INVOKEVIRTUAL net/minecraft/world/lighting/WorldLightManager.getLightEngine (Lnet/minecraft/world/LightType;)Lnet/minecraft/world/lighting/IWorldLightListener;
            ALOAD 2
            INVOKEINTERFACE net/minecraft/world/lighting/IWorldLightListener.getLightFor (Lnet/minecraft/util/math/BlockPos;)I (itf)
            ISTORE 3
           L1
            LINENUMBER 113 L1
            ALOAD 0
            ALOAD 2
            ILOAD 3
            INVOKESTATIC atomicstryker/dynamiclights/client/DynamicLights.getDynamicLightValue (Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;I)I
            IRETURN
           L2
            LOCALVARIABLE this Latomicstryker/dynamiclights/client/WorldTest; L0 L2 0
            LOCALVARIABLE type Lnet/minecraft/world/LightType; L0 L2 1
            LOCALVARIABLE pos Lnet/minecraft/util/math/BlockPos; L0 L2 2
            LOCALVARIABLE i I L1 L2 3
            MAXSTACK = 3
            MAXLOCALS = 4
        }

     */

    return {
        'Dynamic Lights Hook': {
            'target': {
                'type': 'CLASS',
                'name': className
            },
            'transformer': function (classNode) {

                print("dynamic lights coremod visiting class ", classNode.name);

                var methods = classNode.methods;
                for (i = 0; i < methods.length; i++) {
                    print("dynamic lights coremod looking at method ", methods[i].name, "desc: ", methods[i].desc);
                    if (methods[i].name.equals(methodName) && methods[i].desc.equals(methodDescriptorToModify)) {
                        print("this is our target method!");

                        /* inject this before IRETURN:

                            ISTORE 3
                            ALOAD 0
                            ALOAD 2
                            ILOAD 3
                            INVOKESTATIC atomicstryker/dynamiclights/client/DynamicLights.getDynamicLightValue (Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;I)I

                         */

                        var targetNode;
                        var iter = methods[i].instructions.iterator();
                        var index = 0;
                        var found = false;
                        while (iter.hasNext()) {
                            // check all nodes
                            targetNode = iter.next();

                            // find the IRETURN node
                            if (targetNode.getOpcode() === Opcodes.IRETURN) {
                                print("Found IRETURN Node at index ", index);
                                found = true;
                                break;
                            }
                            index++;
                        }
                        if (found) {

                            var toInject = new InsnList();
                            toInject.add(new VarInsnNode(Opcodes.ISTORE, 3));
                            toInject.add(new VarInsnNode(Opcodes.ALOAD, 0));
                            toInject.add(new VarInsnNode(Opcodes.ALOAD, 2));
                            toInject.add(new VarInsnNode(Opcodes.ILOAD, 3));
                            toInject.add(new MethodInsnNode(Opcodes.INVOKESTATIC, "atomicstryker/dynamiclights/client/DynamicLights", "getDynamicLightValue", goalInvokeDesc, false));

                            // now write our instructions before the target node
                            methods[i].instructions.insertBefore(targetNode, toInject);
                            print("Dynamic Lights injected code!");
                        }
                        break;
                    }
                }
                print("Dynamic Lights coremod exiting");
                return classNode;
            }
        }
    }
}